File: gitdot-web/packages/api/src/endpoint/user/list-user-commits.ts
import { z } from "zod";
import { page, UserCommitResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const ListUserCommitsRequest = z.object({
  from: z.iso.datetime().optional(),
  to: z.iso.datetime().optional(),
  cursor: z.string().optional(),
  limit: z.number().int().positive().optional(),
});
export type ListUserCommitsRequest = z.infer<typeof ListUserCommitsRequest>;

export const ListUserCommitsResponse = page(UserCommitResource);
export type ListUserCommitsResponse = z.infer<typeof ListUserCommitsResponse>;

export const ListUserCommits = {
  path: "/user/{user_name}/commits",
  method: "GET",
  request: ListUserCommitsRequest,
  response: ListUserCommitsResponse,
} as const satisfies Endpoint;
export type ListUserCommits = typeof ListUserCommits;

File: gitdot-api/src/endpoint/user/list_user_commits.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::{
    endpoint::Endpoint,
    resource::{common::Page, user::UserCommitResource},
};

pub struct ListUserCommits;

impl Endpoint for ListUserCommits {
    const PATH: &'static str = "/user/{user_name}/commits";
    const METHOD: http::Method = http::Method::GET;

    type Request = ListUserCommitsRequest;
    type Response = ListUserCommitsResponse;
}

#[derive(ApiRequest, Debug, Default, Serialize, Deserialize)]
pub struct ListUserCommitsRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub from: Option<DateTime<Utc>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub to: Option<DateTime<Utc>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cursor: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
}

pub type ListUserCommitsResponse = Page<UserCommitResource>;
